In [1]:
import pandas as pd
import plotly.express as px
import plotly.graph_objects as go
import random 
import numpy as np
import matplotlib.pyplot as plt

In [56]:
pokemons = pd.read_csv('boss1.csv')
pokemons.head()

Unnamed: 0.1  Unnamed: 0  pokedex_number        name german_name  \
0             0           0               1   Bulbasaur     Bisasam   
1             1           1               2     Ivysaur   Bisaknosp   
2             2           2               3    Venusaur    Bisaflor   
3             4           4               4  Charmander    Glumanda   
4             5           5               5  Charmeleon     Glutexo   

         japanese_name  generation  status         species  type_number  ...  \
0  フシギダネ (Fushigidane)           1  Normal    Seed Pokémon            2  ...   
1   フシギソウ (Fushigisou)           1  Normal    Seed Pokémon            2  ...   
2  フシギバナ (Fushigibana)           1  Normal    Seed Pokémon            2  ...   
3      ヒトカゲ (Hitokage)           1  Normal  Lizard Pokémon            1  ...   
4       リザード (Lizardo)           1  Normal   Flame Pokémon            1  ...   

  against_ground against_flying  against_psychic  against_bug  against_rock  \
0            1.0            2.0              2.0          1.0           1.0   
1            1.0            2.0              2.0          1.0           1.0   
2            1.0            2.0              2.0          1.0           1.0   
3            2.0            1.0              1.0          0.5           2.0   
4            2.0            1.0              1.0          0.5           2.0   

  against_ghost against_dragon against_dark  against_steel  against_fairy  
0           1.0            1.0          1.0            1.0            0.5  
1           1.0            1.0          1.0            1.0            0.5  
2           1.0            1.0          1.0            1.0            0.5  
3           1.0            1.0          1.0            0.5            0.5  
4           1.0            1.0          1.0            0.5            0.5  

[5 rows x 52 columns]

In [57]:
pokemons.value_counts('generation')

generation
1    172
5    150
3    130
4    100
2     94
8     82
6     75
7     69
Name: count, dtype: int64

In [58]:
pokemons_by_generation = pokemons.groupby("generation").agg({"attack": "mean", "speed": "mean", "defense": "mean"})
pokemons_by_generation["attack"]

fig = go.Figure()
fig.add_trace(go.Histogram(x=pokemons["generation"], nbinsx=15, name="Quantidade", marker_color="#76b5c5"))

fig.add_trace(go.Scatter(x=pokemons_by_generation.index, y=pokemons_by_generation["attack"],
                    mode='lines+markers',
                    name='Ataque'))
fig.add_trace(go.Scatter(x=pokemons_by_generation.index, y=pokemons_by_generation["speed"],
                    mode='lines+markers',
                    name='Velocidade',
                    marker_color="green"))
fig.add_trace(go.Scatter(x=pokemons_by_generation.index, y=pokemons_by_generation["defense"],
                    mode='lines+markers',
                    name='Defesa',
                    marker_color="yellow"))

fig.show()

- Análise: A média do ataque teve uma tendência de crescimento a partir da segunda geração, e teve o ápice na quinta geração, em seguinda teve uma queda na sexta e voltou bruscamente na sétima.
A segunda e sexta geração são as únicas cuja defesa média ultrapassou o ataque.


In [59]:
#Boxsplot
pokeataque = px.box(pokemons, x="type_1", y="attack", points="all", color='type_1',hover_data='name', labels={"type_1":"Tipo 1", "attack":"Ataque!",})
pokeataque.show()

pokespeed = px.box(pokemons, x="type_1", y="speed", points="all", color='type_1',hover_data='name', labels={"type_1":"Tipo 1", "speed":"Velocidade!",})
pokespeed.show()

pokedefense = px.box(pokemons, x="type_1", y="defense", points="all", color='type_1',hover_data='name',labels={"type_1":"Tipo 1", "defense":"Defesa!",})
pokedefense.show()



## Análise:
Em minha analise notei que o pokemon do tipo pedra Rampardos, com um ataque de 160 pontos, em seguida verifiquei os mais fracos  onde encontrei dois com a mesma pontuação seriam eles o Happiny e a Chansey ambos com 5 de ataque do tipo Normal.

A respeito da velocidade o pokemon Ninjask se destaca com uma velocidade de 160d o tipo Inseto, e o oposto dele vemos  3 tipos de pokemons, o Pyukumunuku, Shuckle e o Munchlax respectivamente dos tipos Água, Inseto e Normal.

Por último temos a defesa onde curiosamente do tipo Inseto, o Shuckle se destaca  com  incriveis 230 pontos de defesa, o que idenficamos como um outlier pois está distante da média da defesa dos pokemons que seria 60 pontos e o menos defensivo que idenficamos são Chansey e Happiny do tipo Normal.

### Existe algum tipo notavelmente equilibrado ou forte em algum atributo?
Na defesa percebos que o tipo Steel lidera, quando falamos sobre equilibrio o tipo Water se destaca.

In [60]:
#Comparação de Ataque  do tipo com todos os pokemons.
pokemonmediaA = pokemons.groupby('type_1')['attack'].mean()
print(pokemonmediaA)

pokemonmediaallA = pokemons['attack'].mean()
print(f'média  de attack de todos pokemons {pokemonmediaallA}')

#Comparação de Velocidade  do tipo com todos os pokemons.
pokemonmediaV = pokemons.groupby('type_1')['speed'].mean()
print(pokemonmediaA)

pokemonmediaallV = pokemons['speed'].mean()
print(f'média  de speed de todos pokemons {pokemonmediaallV}')
#Comparação de Defesa  do tipo com todos os pokemons.
pokemonmediaD = pokemons.groupby('type_1')['defense'].mean()
print(pokemonmediaD)

pokemonmediaallD = pokemons['defense'].mean()
print(f'média  de defesa de todos pokemons {pokemonmediaallD}')

type_1
Bug         63.797297
Dark        76.763158
Dragon      93.173913
Electric    70.538462
Fairy       56.157895
Fighting    97.515152
Fire        79.363636
Flying      64.333333
Ghost       70.054054
Grass       70.487805
Ground      88.081081
Ice         75.878788
Normal      72.180952
Poison      74.285714
Psychic     52.723404
Rock        87.211538
Steel       86.791667
Water       71.516667
Name: attack, dtype: float64
média  de attack de todos pokemons 73.94151376146789
type_1
Bug         63.797297
Dark        76.763158
Dragon      93.173913
Electric    70.538462
Fairy       56.157895
Fighting    97.515152
Fire        79.363636
Flying      64.333333
Ghost       70.054054
Grass       70.487805
Ground      88.081081
Ice         75.878788
Normal      72.180952
Poison      74.285714
Psychic     52.723404
Rock        87.211538
Steel       86.791667
Water       71.516667
Name: attack, dtype: float64
média  de speed de todos pokemons 63.684633027522935
type_1
Bug          69.324324


In [61]:
unique_types = pokemons['type_1'].unique()

selected_types = random.sample(list(unique_types), 10)

selected_pokemons = pd.DataFrame()

for p_type in selected_types:
    type_pokemons = pokemons[pokemons['type_1'] == p_type]
    selected_pokemons = pd.concat([selected_pokemons, type_pokemons.sample(3)], ignore_index=True)

pesos = {
 'speed': 2.0,
 'defense': 1.0,
 'attack': 2.0,
}

selected_pokemons['metrica'] = selected_pokemons['attack'] * pesos['attack']  + selected_pokemons['defense'] * pesos['defense'] + selected_pokemons['speed'] * pesos['speed']

selected_pokemons = selected_pokemons.rename(columns={'against_fight': 'against_fighting'})

for index, row in selected_pokemons.iterrows():
    for type in selected_pokemons['type_1'].unique():
        selected_pokemons.loc[index,'metrica'] += row[f'against_{type.lower()}']

    selected_pokemons.loc[index,'metrica'] = row['metrica'] / 23


selected_list_pokemons = []
for type in selected_pokemons['type_1'].unique():
    selected_type = selected_pokemons[selected_pokemons['type_1'] == type]
    selected_list_pokemons.append(selected_pokemons.iloc[selected_type['metrica'].idxmax()])


current_team_pokemon = pd.DataFrame(selected_list_pokemons)

current_team_pokemon = current_team_pokemon.sort_values(by="metrica", ascending=False).head(6)

current_team_pokemon.drop(columns=['Unnamed: 0.1','Unnamed: 0','german_name','japanese_name' ])



pokedex_number                      name  generation  status  \
25             555  Darmanitan Standard Mode           5  Normal   
22             149                 Dragonite           1  Normal   
15             448                   Lucario           4  Normal   
14             652                Chesnaught           6  Normal   
9              781                  Dhelmise           7  Normal   
6              823               Corviknight           8  Normal   

                species  type_number    type_1    type_2  height_m  weight_kg  \
25      Blazing Pokémon            1      Fire       NaN       1.3       92.9   
22       Dragon Pokémon            2    Dragon    Flying       2.2      210.0   
15         Aura Pokémon            2  Fighting     Steel       1.2       54.0   
14  Spiny Armor Pokémon            2     Grass  Fighting       1.6       90.0   
9   Sea Creeper Pokémon            2     Ghost     Grass       3.9      210.0   
6         Raven Pokémon            2    Flying     Steel       2.2       75.0   

    ...  against_flying against_psychic against_bug against_rock  \
25  ...             1.0             1.0        0.50         2.00   
22  ...             1.0             1.0        0.50         2.00   
15  ...             1.0             1.0        0.25         0.25   
14  ...             4.0             2.0        1.00         0.50   
9   ...             2.0             1.0        1.00         1.00   
6   ...             0.5             0.5        0.25         1.00   

    against_ghost  against_dragon  against_dark  against_steel  against_fairy  \
25            1.0             1.0           1.0            0.5            0.5   
22            1.0             2.0           1.0            1.0            2.0   
15            1.0             0.5           0.5            0.5            1.0   
14            1.0             1.0           0.5            1.0            2.0   
9             2.0             1.0           2.0            1.0            1.0   
6             1.0             0.5           1.0            0.5            0.5   

      metrica  
25  22.826087  
22  22.739130  
15  20.434783  
14  20.173913  
9   19.217391  
6   17.956522  

[6 rows x 49 columns]

#### Crie um gráfico de dispersão 3D para visualizar attack, defense e speed dos Pokémon
selecionados.

In [20]:
fig = px.scatter_3d(current_team_pokemon, x = 'attack', y = 'defense', z = 'speed',
                    color = 'name', size_max = 18,
                    title = ('Gráfico de dispersão 3d'))

fig.update_layout(margin=dict(l=0, r=0, b=0, t=0))
fig.show()

###  Justifique a seleção dos 6 Pokémon, explicando como os critérios otimizados ajudaram a montar uma equipe equilibrada. 


#### O nosso time apresenta uma ampla variação de tipos, o que nos ajuda a enfrentar praticamente todos os tipos de oponentes. Além disso, notamos que todos os Pokémon do time possuem um nível de ataque acima de 100, evidenciando que temos um time forte no aspecto ofensivo. As defesas também são sólidas, com todos os membros apresentando valores superiores a 60, o que proporciona uma boa resistência em batalha.

#### Embora a velocidade não seja o ponto forte da equipe, contamos com Unfezant, que possui o maior nível de velocidade. Ele pode ser utilizado para neutralizar alvos distantes rapidamente, caso necessário. Por outro lado, Doublade, com uma defesa impressionante de 150, pode facilmente aguentar ataques, protegendo os membros mais frágeis do time.

#### Se optarmos por combinar os ataques, isso poderia resultar em um desempenho explosivo, dado o alto nível de ataque de nossa equipe. Os Pokémon mais robustos podem absorver o dano, enquanto os mais rápidos se encarregam de eliminar os adversários com agilidade.

## Discuta os trade-offs entre atributos ofensivos e defensivos, além das resistências e fraquezas de cada Pokémon contra outros tipos. 

### Excadrill (Terra/Aço)
#### HP: 110
#### Ataque: 135
#### Defesa: 60
#### Velocidade: 88
#### Bom Contra: Pedra (2.0), Fada (2.0), Elétrico (2.0)
#### Ruim Contra: Água (0.5), Terra (0.5)
### Unfezant (Normal/Voador)
#### HP: 80
#### Ataque: 115
#### Defesa: 80
#### Velocidade: 93
#### Bom Contra: Inseto (2.0), Grama (2.0), Luta (2.0)
#### Ruim Contra: Pedra (0.5), Fantasma (0.5)
### Kabutops (Pedra/Água)
#### HP: 60
#### Ataque: 115
#### Defesa: 100
#### Velocidade: 80
#### Bom Contra: Fogo (2.0), Voador (2.0), Terra (2.0)
#### Ruim Contra: Lutador (0.5), Elétrico (0.5)
### Galarian Rapidash (Psíquico/Fada)
#### HP: 65
#### Ataque: 100
#### Defesa: 70
#### Velocidade: 105
#### Bom Contra: Lutador (2.0), Dragão (2.0), Sombrio (2.0)
#### Ruim Contra: Aço (0.5), Venenoso (0.5)
### Luxray (Elétrico)
#### HP: 80
#### Ataque: 120
#### Defesa: 79
#### Velocidade: 70
#### Bom Contra: Água (2.0), Voador (2.0)
#### Ruim Contra: Terra (0.5), Elétrico (0.5)
### Doublade (Aço/Fantasma)
#### HP: 59
#### Ataque: 110
#### Defesa: 150
#### Velocidade: 35
#### Bom Contra: Fada (2.0), Gelo (2.0)
#### Ruim Contra: Lutador (0.5), Fada (0.5)

In [63]:
selected_pokemons_against = current_team_pokemon[['against_normal', 'against_fire', 'against_water', 'against_electric',
       'against_grass', 'against_ice', 'against_fighting', 'against_poison',
       'against_ground', 'against_flying', 'against_psychic', 'against_bug',
       'against_rock', 'against_ghost', 'against_dragon', 'against_dark',
       'against_steel', 'against_fairy']]

for index,value in enumerate(list(selected_pokemons_against.values)):
    fig = px.line_polar(selected_pokemons_against, r=value, theta=list(selected_pokemons_against.columns), line_close=True, title=current_team_pokemon['name'].iloc[index])
    fig.update_traces(fill='toself')  
    fig.show()


In [64]:
fig = px.imshow(
    selected_pokemons_against,
    labels=dict(x="type_1", y="name"),
    x=selected_pokemons_against.columns,
    y=selected_pokemons_against.index,
    color_continuous_scale="Viridis",
)

fig = px.imshow(list(selected_pokemons_against.values), y=list(current_team_pokemon["name"].values), labels=dict(x="type_1", y="name"), x=list(selected_pokemons_against.columns), color_continuous_scale="Blackbody")
fig.show()